stdlib/public/runtime/SymbolInfo.cpp

```diff
@@ -17,27 +17,81 @@
 #define NOMINMAX
 #include <Windows.h>
 #include <DbgHelp.h>
+#include <Psapi.h>
+#include <shlwapi.h>
 #elif SWIFT_STDLIB_HAS_DLADDR
 #include <dlfcn.h>
 #endif
 
-#include "swift/Runtime/Win32.h"
+#include <array>
+#include <cstdlib>
 
 #include "ImageInspection.h"
+#include "swift/Runtime/Win32.h"
 
 using namespace swift;
 
-const char *SymbolInfo::getFilename() const {
+SymbolInfo::SymbolInfo()
 #if defined(_WIN32) && !defined(__CYGWIN__)
-  return nullptr;
+: _package({}),
+#elif SWIFT_STDLIB_HAS_DLADDR
+: _info({}),
+#endif
+_address(nullptr)
+{
+}
+
+SymbolInfo::~SymbolInfo() {
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  if (char *imagePath = _imagePath.getValueOr(nullptr)) {
+    free(imagePath);
+  }
+#endif
+}
+
+const void *SymbolInfo::getAddress() const {
+  return _address;
+}
+
+const char *SymbolInfo::getImagePath() const {
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  if (!_imagePath.has_value()) {
+    // Call GetMappedFileNameW() instead of SymGetModuleInfoW() because it
+    // doesn't require the use of the global DbgHelp lock and doesn't fetch
+    // other information we don't need.
+    std::array<wchar_t, MAX_PATH> filename;
+    if (GetMappedFileNameW(GetCurrentProcess(), const_cast<void *>(_address),
+                           filename.data(), filename.size())) {
+      _imagePath = _swift_win32_copyUTF8FromWide(filename.data());
+    }
+  }
+
+  return _imagePath.getValueOr(nullptr);
 #elif SWIFT_STDLIB_HAS_DLADDR
   return _info.dli_fname;
 #else
   return nullptr;
 #endif
 }
 
-const void *SymbolInfo::getBaseAddress() const {
+const char *SymbolInfo::getImageFilename() const {
+  // TODO: if we adopt C++17, consider using std::filesystem::path::filename()
+  const char *imagePath = getImagePath();
+  if (!imagePath) {
+    return nullptr;
+  }
+#if defined(_WIN32) && !defined(__CYGWIN__)
+  return PathFindFileNameA(imagePath);
```

Comment:
I worry about this as _technically_ `pszImagePath` is limited to `MAX_PATH`, which may or may not hold, particularly in the Swift world where we go through efforts to use NT style paths to avoid the length limitations.  A far less elegant solution which may be safer (I haven't thought hard enough about it) is to use `PathCchRemoveFileSpec` and then use that to compute the offset into the buffer to compute the "file spec" (why Microsoft does not provide a `PathCchGetFileSpec` is unclear to me).

Reply:
Indeed, when it says that's the limit, it's probably not kidding; most likely `PathFindFileNameA` has a buffer of that size that it uses to convert from ANSI to UTF-16.

Which also brings up the related point that this function will assume ANSI encoding and might produce unusual results if passed UTF-8, which is what we currently have.

Reply:
Easiest would be to just find the last slash and truncate, same as what we do on Darwin/Linux.